dimod/response.py

```diff
@@ -13,737 +13,469 @@
-                Vartype of the response. If not provided, vartype is inferred from the
-                samples matrix if possible or a ValueError is raised.
+            fields (list, optional, default=None):
+                If specified, only these fields' values are included in the yielded tuples.
+                The special field name 'sample' can be used to view the samples.
 
-            info (dict, optional, default=None):
-                Information about the response as a whole formatted as a dict.
+            sorted_by (str/None, optional, default='energy'):
+                Selects the record field used to sort the samples. If None, the samples are yielded
+                in record order.
 
-        Returns:
-            :obj:`.Response`: A `dimod` :obj:`.Response` object based on the input dicts.
+            name (str/None, optional, default='Sample'):
+                Name of the yielded namedtuples or None to yield regular tuples.
 
-        Raises:
-            :exc:`ValueError`: If vartype is not provided and samples are all 1s, have more
-                than two unique values, or have values of an unknown vartype.
+        Yields:
+            namedtuple/tuple: The data in the response, in the order specified by the input
+            `fields`.
 
         Examples:
-            This example code snippet builds a response from an interable of samples and
-            its corresponding dict of energies.
 
-            .. code-block:: python
-
-                samples = [{'a': -1, 'b': +1}, {'a': +1, 'b': -1}]
-                energies = [-1.0, -1.0]
-                response = Response.from_dicts(samples, {'energy': energies})
+            >>> import dimod
+            ...
+            >>> response = dimod.ExactSolver().sample_ising({'a': -0.5, 'b': 1.0}, {('a', 'b'): -1})
+            >>> for datum in response.data():   # doctest: +SKIP
+            ...     print(datum)
+            Sample(sample={'a': -1, 'b': -1}, energy=-1.5)
+            Sample(sample={'a': 1, 'b': -1}, energy=-0.5)
+            Sample(sample={'a': 1, 'b': 1}, energy=-0.5)
+            Sample(sample={'a': -1, 'b': 1}, energy=2.5)
+            >>> for energy, in response.data(fields=['energy'], sorted_by='energy'):
+            ...     print(energy)
+            ...
+            -1.5
+            -0.5
+            -0.5
+            2.5
+            >>> print(next(response.data(fields=['energy'], name='ExactSolverSample')))
+            ExactSolverSample(energy=-1.5)
 
         """
-        samples = iter(samples)
+        record = self.record
 
-        # get the first sample
-        first_sample = next(samples)
+        if fields is None:
+            # make sure that sample, energy is first
+            fields = ['sample', 'energy', 'num_occurrences'] + [field for field in record.dtype.fields
+                                                                if field not in {'sample', 'energy', 'num_occurrences'}]
 
```

Comment:
Maybe keep a list of these default fields somewhere, class-local, or module-local?